### PR TITLE
Prevent destructive action on production database

### PR DESCRIPTION
**Pull request by schneems**
Prevent destructive action on production database

It is possible to run your tests against your production database by accident right now. While infrequently, but as an anecdotal data point, Heroku receives a non-trivial number of requests for a database restore due to this happening. In these cases the loss can be large. 

To prevent against running tests against production we can store the "environment" version that was used when migrating the database in a new internal table. Before executing tests we can see if the database is a listed in `protected_environments` and abort. There is a manual escape valve to force this check from happening with environment variable `DISABLE_DATABASE_ENVIRONMENT_CHECK=1`

### CODE REVIEW

**Comment by rafaelfranca** [comment]
I think we have to make sure that `rails test` also check this.

File: activerecord/lib/active_record/migration.rb [hunk 1]
```diff
@@ -1013,10 +1033,15 @@ def record_version_state_after_migrating(version)
...
       else
         migrated << version
-        ActiveRecord::SchemaMigration.create!(:version => version.to_s)
+        ActiveRecord::SchemaMigration.create!(version: version.to_s, environment: ActiveRecord::ConnectionHandling::DEFAULT_ENV.call)
       end
     end
 
+    def self.production?
+      raise NoEnvironmentInSchemaError unless SchemaMigration.environment_is_stored?
+      ActiveRecord::SchemaMigration.order(:created_at).last.environment == "production"
```
Comment (sgrif) [inline activerecord/lib/active_record/migration.rb]: Should we just compare to the current environment, rather than special casing production?

File: activerecord/lib/active_record/railtie.rb [hunk 2]
```diff
@@ -36,6 +36,13 @@ class Railtie < Rails::Railtie # :nodoc:
     config.eager_load_namespaces << ActiveRecord
 
     rake_tasks do
+      task :test do
+        if !ENV['RUN_AGAINST_PRODUCTION_DATABASE'] && ActiveRecord::Migrator.production?
```
Comment (sgrif) [inline activerecord/lib/active_record/railtie.rb]: If you accept my previous comment, we should probably rename this to something like `DISABLE_DATABASE_ENVIRONMENT_CHECK`

File: activerecord/lib/active_record/migration.rb [hunk 3]
```diff
@@ -47,6 +47,26 @@ def initialize
...
+        super("#{msg} RAILS_ENV=#{::Rails.env}")
+      else
+        super(msg)
+      end
+    end
+  end
+
+  class ProductionRestrictedError < ActiveRecordError#:nodoc:
+    def initialize
+      msg = "You are attempting to run a destructive action against your 'production' database\n"
```
Comment (sgrif) [inline activerecord/lib/active_record/migration.rb]: Do we want this error message to consider the case where the environment field is blank because they haven't migrated production yet? Is that too much of an edge case?

Comment (schneems) [inline activerecord/lib/active_record/migration.rb]: They should always get `NoEnvironmentInSchemaError` which would be raised by the `production?` method call before this error gets raised when the environment isn't set.

File: activerecord/lib/active_record/migration.rb (hunk 1 above)
Comment (guilleiguaran) [inline activerecord/lib/active_record/migration.rb]: :+1:

Comment (senny) [inline activerecord/lib/active_record/migration.rb]: This won't work. When using `db/structure.sql` you'll get the records from that table in your dumped output like so:

```sql
INSERT INTO schema_migrations (version) VALUES ('20150422072614');

INSERT INTO schema_migrations (version) VALUES ('20150807113927');

INSERT INTO schema_migrations (version) VALUES ('20150811135959');
```

This file is then used to produce a new `_test` database and will have `development` as environment and not `test`.

**Comment by pixeltrix** [comment]
I think what @senny said is fatal flaw for doing it via a column in the `schema_migrations` table and I wasn't a fan of encoding it there anyway. What about creating a new `schema_environment` table that just has a single row and then don't add the value to `db/structure.sql`?

**Comment by schneems** [comment]
Putting this data in it's own table and keeping it out of schema/structure on disk seems like a good idea, I'll take a stab in a bit. I've been busy with sprockets recently. Thanks for the preliminary review everyone.

**Comment by schneems** [comment]
I found out that the schema_migration table is already ignored in the "schema" dumper https://github.com/rails/rails/blob/4d77e02d344349e899d6c4c5e10b1879867b72eb/activerecord/lib/active_record/schema_dumper.rb#L250.

Unfortunately the structure dumps are all done via individual database adapters such as https://github.com/rails/rails/blob/4d77e02d344349e899d6c4c5e10b1879867b72eb/activerecord/lib/active_record/tasks/postgresql_database_tasks.rb#L45-L66.

So we've got two options. 

## Option 1

We can find a way to exclude a table from the dump in mysql, postgresql, and sqlite3 (the only 3 database tasks in active record). It looks like you can do this in postgres with `--exclude-schema=schema` [docs](http://www.postgresql.org/docs/9.3/static/app-pgdump.html).  Then we'll have to find and modify the other popular adapters such as oracle, and sql server. Then we have to repeat the process for jruby drivers like `activerecord-jdbc-adapter` and if all of them are modified correctly, then things will work.

- **Cons:** It's a bunch of work. Different adaptors responsible for testing and not regressing.
- **Pros:** When it works, everything works there's no caveats.

## Option 2

Instead of modifying the dump, we modify our database tasks so that a `structure:load` will also update the database with the current environment. 

- **Cons:** If you're dumping the structure and then using the psql command to import it instead of a database task you would still end up with this data in your environment. I don't know how many people are doing this. Even if someone does this, they're back to where we are now, there's no regression they just won't be protected. Everyone not doing this would be protected

- **Pros:** It's way easier to implement.

What do you think?

On an aside, can someone explain the differences between schema/structure dumping & loading. Why would I want to use one versus the other?

**Comment by al2o3cr** [comment]
@schneems Structure dumping is a requirement if the database contains things that can't be correctly dumped in Ruby format: views, database-specific types (MySQL enums, sets), database-specific information like column comments, foreign key constraints prior to AR 4.2, etc.

**Comment by al2o3cr** [comment]
On a general note, while I like the premise here I think it's just removing the most obvious footguns. Running locally in development mode with `DATABASE_URL` set to production has plenty of additional hazards both from the framework (`db:drop`, anyone?) and the user's own code.

**Comment by bilbosan** [comment]
Guys,

It isn't just production databases. I had an issue today where I wiped my development database because of this. That isn't as catastrophic, but still counts as a major pain. Aside from the fact that it took me a while to figure out what was going on and why all my test factories no longer worked.

In my opinion, the config/database.yml entries for test should always win.

**Comment by schneems** [comment]
Finally got all tests passing. I'm now using a new table that we are storing environment data in called `environment_checks` using the corresponding `EnvironmentCheck` model.

We are scrubbing it from the `schema.rb` and manually adding it back in and then populating the value. We are not also manually adding it back into after at test schema load and populating the value.

The "protected" environments are defined as an array so you could have `"production"` and `"staging"` at the same time, you could even add development in there if you wanted.

If you either don't have a database (cannot connect to the db) or there is nothing in your database (schema version is 0) then we don't need to protect against losing data. This behavior eliminated the need to add a bunch of edge case logic to a large number of tests, I think it's a fine assumption.


I added in a database deletion guard against `db:test:purge` which is called by `db:test:prepare` which is the recommended way to prepare databases for tests: http://guides.rubyonrails.org/testing.html#maintaining-the-test-database-schema via the guides.

We are also guarding all other destructive database rake tasks including:

- db:drop:all
- db:drop
- db:purge:all
- db:purge

Now if anyone wants to safeguard a rake task they can have it depend on `db:check_protected_environments`. The biggest gem we probably want to add a check to would be `database_cleaner`. The test is in a method on DatabaseTasks, so it can be called even if you're not using Rake.

I had to add in a `db:purge:_unsafe` task and change a test to use it. The test was ensuring that we would get an error message when purging a non-existant database, however the test uses sqlite3 which will create a database when you try to connect to it. The EnvironmentCheck model would create a database connection in order to see if the table had been created yet, when it did it sqlite3 would create the database.

There is one other hiccup with the tests: mysql doesn't store fractional seconds by default and the table gets updated faster than a second, I had to add a sleep into one of the tests to get it consistently passing. That won't be an issue in real life unless people are running multiple migrations within fractional seconds of each other with different environments.

I did want to comment on the behavior @senny described. cc/ @pixeltrix the described behavior of loading schema versions is done manually: https://github.com/rails/rails/blob/7f86ed8b36b534cf94fe2ff47ae71956b57893ac/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L958-L964 we are selectively only inserting the version and nothing else. The original solution of storing the data in the SchemaMigration table would have been fine unless we also explicitly added code that would put the environment fields in there as well. Splitting it out into another table makes things a little more explicit and keeps logic separate which i'm fine with.

**Comment by schneems** [comment]
Ping, thoughts? Would love to get this into the beta before RC.

**Comment by rafaelfranca** [comment]
:eyes: now!

File: activerecord/lib/active_record/environment_check.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      # Creates a schema table with columns +environment+ and +version+
+      def create_table
+        if table_exists?
+          return if environment_is_stored?
+          connection.change_table(table_name) do |t|
+            t.add_column :environment, :string
+            t.timestamps
+          end
+        else
+
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/environment_check.rb]: :scissors:

File: activerecord/lib/active_record/environment_check.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      end
+
+      def table_exists?
+        ActiveSupport::Deprecation.silence { connection.table_exists?(table_name) }
+      end
+
+      # Creates a schema table with columns +environment+ and +version+
+      def create_table
+        if table_exists?
+          return if environment_is_stored?
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/environment_check.rb]: a blank line here would made explicit the early return

File: activerecord/lib/active_record/schema_dumper.rb [hunk 6]
```diff
@@ -254,7 +254,7 @@ def remove_prefix_and_suffix(table)
       end
 
       def ignored?(table_name)
-        [ActiveRecord::Base.schema_migrations_table_name, ignore_tables].flatten.any? do |ignored|
+        [ActiveRecord::Base.schema_migrations_table_name, ActiveRecord::Base.environment_check_table_name , ignore_tables].flatten.any? do |ignored|
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/schema_dumper.rb]: there is a space before the `,`

**Comment by rafaelfranca** [comment]
LGTM :shipit:

**Comment by matthewd** [comment]
I'm not a fan of the database pollution this creates... should there be a way to opt out of it, at least?

**Comment by schneems** [comment]
@matthewd You can set `config.protected_environments =[]` to avoid the check from running, but it won't stop your database from being updated. Can you be more specific about your concerns with  database pollution? We aren't storing coal ash. This PR adds roughly the same data to the database as SchemaMigration data and most people aren't even aware that is in their DB. If you're worried about too many entries we could write logic to check to see if the last environment was the same before updating the DB, that should be easy enough and if you're always on the same system means it should only have one row. If you're worried about adding another table we could fold this into the schema_migration table which was actually my first implementation. I like that the two are independent for clarity sake, but i'm not tied to the implementation. I don't think it's very maintainable to wrap every bit of this PR in a conditional. If that's the route we want to go down, we should make schema migration information in the DB optional as well.

File: activerecord/lib/active_record/environment_check.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+          connection.change_table(table_name) do |t|
+            t.add_column :environment, :string
```
Comment (matthewd) [inline activerecord/lib/active_record/environment_check.rb]: When does this apply?

Comment (schneems) [inline activerecord/lib/active_record/environment_check.rb]: It is called 3 different ways. When schema is loaded into a database `rake db:schema:load`, when a structure is loaded `rake db:structure:load`, and when anything is done with migrations like `rake db:migrate`.

Comment (matthewd) [inline activerecord/lib/active_record/environment_check.rb]: Right.. but I don't see when the table would exist without the column.

Comment (schneems) [inline activerecord/lib/active_record/environment_check.rb]: Ahh, I understand now. Good catch, i don't think it can be. I think this is a fragment from when I was updating the SchemaMigration table instead of using a seperate environment check table. I'm updating to remove that logic.

Comment (schneems) [inline activerecord/lib/active_record/environment_check.rb]: Fixed

**Comment by prathamesh-sonpatki** [comment]
@schneems This can be closed now?
